### PR TITLE
Registering WC Pay with Nav clientside

### PR DESCRIPTION
**Pull request by psealock**
While reviewing https://github.com/Automattic/woocommerce-payments/pull/1067 I noticed a big flaw in how we register items on the client. We were trying to use the `path` as a unique identifier which failed when items were being registered to a `parent` because a mismatch in `id`. More on that in https://github.com/woocommerce/woocommerce-admin/pull/5773

This PR updates to the changes introduced in https://github.com/woocommerce/woocommerce-admin/pull/5773

#### Testing instructions

Use the WC Admin branch in https://github.com/woocommerce/woocommerce-admin/pull/5773 to make sure links work correctly and do not incur a page refresh when they should not. Specifically, these issues should be resolved: https://github.com/Automattic/woocommerce-payments/pull/1067#issuecomment-735455853

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

**Comment by DanReyLop** [comment]
It works! 🎉  The Deposits, Transactions, and Disputes pages don't trigger a full-page reload now.

However, this is still happening:
> When clicking on the `WooCommerce Payments > Settings` menu item, the correct page is loaded (`/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments`), but the navigation sidebar goes completely black.

Screenshot 2020-12-01 at 12 46 15

File: client/index.js
Comment (joshuatf) [inline client/index.js:34]: This will replace the `WooCommerce Payments` menu with `Connect` under the extensions menu.

Screen Shot 2020-12-01 at 11 27 08 AM

**Review by joshuatf — COMMENTED** [review]
Thanks for opening up this fix, Paul!

Using the latest `main` branch of WCA, I'm not able to see any items which appears to be the same mismatch in expected ID from `wc-payments` to `wc_admin-wc-admin&path=/payments/deposits`.

While I think client-side `navArgs` might be useful, I'm also a little concerned that we lose some of the benefits we discussed before of WCA registration if we have to redeclare IDs in multiple places.

Should we lean into https://github.com/woocommerce/woocommerce-admin/pull/5665 a bit more and use this auto-generated ID in WC Pay?  Or should we include the updated ID in `navArgs` in https://github.com/Automattic/woocommerce-payments/pull/1067?

**Comment by psealock** [comment]
@DanReyLop interesting, I didn't come across this in my testing. Any errors on the console?

@joshuatf 

>Using the latest main branch of WCA, I'm not able to see any items which appears to be the same mismatch in expected ID from wc-payments to wc_admin-wc-admin&path=/payments/deposits.

Test out https://github.com/Automattic/woocommerce-payments/pull/1067 with the associated wc-admin branch and you'll see no child items appear. They are all mismatched.

>Should we lean into woocommerce/woocommerce-admin#5665 a bit more and use this auto-generated ID in WC Pay?

We could. For extensions such as this one that add items to a parent they have also created I see two options:

1. Ask extensions to make the parent's id. For instance, here the parent would be  `wc_admin-wc-admin&path=/payments/deposits` and not `wc-payments`.

2. We handle the conversion for them. I feel this gets messy, but is possible. It feels like we're making something more and more complicated in order to avoiding an inconvenience. I know I'm backtracking, but it feels more straightforward to ask extensions to supply the id since they need to amend PHP anyways.

**Comment by DanReyLop** [comment]
> Any errors on the console?

```
utils.js:13 Uncaught TypeError: Cannot read property '__' of undefined
    at Module../node_modules/@wordpress/components/build-module/alignment-matrix-control/utils.js (utils.js:13)
    at __webpack_require__ (bootstrap:19)
    at Module../node_modules/@wordpress/components/build-module/alignment-matrix-control/cell.js (cell.js:1)
    at __webpack_require__ (bootstrap:19)
    at Module../node_modules/@wordpress/components/build-module/alignment-matrix-control/index.js (index.js:1)
    at __webpack_require__ (bootstrap:19)
    at Module../node_modules/@wordpress/components/build-module/index.js (index.js:1)
    at __webpack_require__ (bootstrap:19)
    at Module../packages/navigation/build-module/index.js (history.js:66)
    at __webpack_require__ (bootstrap:19)
./node_modules/@wordpress/components/build-module/alignment-matrix-control/utils.js @ utils.js:13
__webpack_require__ @ bootstrap:19
./node_modules/@wordpress/components/build-module/alignment-matrix-control/cell.js @ cell.js:1
__webpack_require__ @ bootstrap:19
./node_modules/@wordpress/components/build-module/alignment-matrix-control/index.js @ index.js:1
__webpack_require__ @ bootstrap:19
./node_modules/@wordpress/components/build-module/index.js @ index.js:1
__webpack_require__ @ bootstrap:19
./packages/navigation/build-module/index.js @ history.js:66
__webpack_require__ @ bootstrap:19
(anonymous) @ bootstrap:83
(anonymous) @ bootstrap:83
settings.js:47 Uncaught TypeError: Cannot read property 'getQuery' of undefined
    at isWCPaySettingsPage (settings.js:47)
    at Module../client/settings.js (settings.js:18)
    at __webpack_require__ (bootstrap:19)
    at bootstrap:83
    at bootstrap:83
isWCPaySettingsPage @ settings.js:47
./client/settings.js @ settings.js:18
__webpack_require__ @ bootstrap:19
(anonymous) @ bootstrap:83
(anonymous) @ bootstrap:83
index.js:375 Uncaught TypeError: Cannot read property 'getHistory' of undefined
    at Module../client/header/activity-panel/index.js (index.js:375)
    at __webpack_require__ (bootstrap:80)
    at Module../client/header/index.js (index.js?ver=1606681983:2295)
    at __webpack_require__ (bootstrap:80)
    at Module../client/layout/index.js (index.js?ver=1606681983:2930)
    at __webpack_require__ (bootstrap:80)
    at Module.<anonymous> (index.js?ver=1606681983:2505)
    at Module../client/index.js (index.js?ver=1606681983:2577)
    at __webpack_require__ (bootstrap:80)
    at bootstrap:260
```
([Screenshot, if you see them better that way](https://user-images.githubusercontent.com/1715800/100794529-9a24e980-3415-11eb-8e4a-68f5aff25ea2.png))

I'm using WooCommerce 4.6.1, and WordPress 5.3.6 (a bit old, but still should be supported).

**Comment by joshuatf** [comment]
>Test out #1067 with the associated wc-admin branch and you'll see no child items appear. They are all mismatched.

I think we might not be on the same page.  I don't see them working on main, I _also_ don't see them working here.  This is supposed to resolve them so that the child items appear correct?

I'm still seeing the parent ID registered on the server as `wc_admin-wc-admin&path=/payments/deposits` while the child items set a parent to `wc-payments`.

>It feels like we're making something more and more complicated in order to avoiding an inconvenience. 

I see what you mean.  It seems like there should be an easier way to handle this, but maybe the way you've done it here is best.

The console errors that @DanReyLop noted look very familiar.  Dependency injection related?

**Comment by psealock** [comment]
> I'm still seeing the parent ID registered on the server as wc_admin-wc-admin&path=/payments/deposits while the child items set a parent to wc-payments.

You'll need to test this out with https://github.com/woocommerce/woocommerce-admin/pull/5773 which is the corresponding PR on wc-admin.

**Comment by psealock** [comment]
Thanks @DanReyLop that is definitely Dependency related. We aren't using Dependency Extraction Webpack Plugin so that could be the problem. This will be fixed in https://github.com/woocommerce/woocommerce-admin/pull/5762.

I can't reproduce the error, but if you switch to this [woo-flavoured Dependency Extraction Webpack Plugin](https://github.com/woocommerce/woocommerce-admin/tree/main/packages/dependency-extraction-webpack-plugin) does it help? It will handle WC dependencies as well as WP deps. I made that change in https://github.com/Automattic/woocommerce-payments/pull/1136/commits/431386cd7020d67854bb097684c6c3e78612776a. Does it resolve the issue? If not, it is likely on our end and will be fixed in https://github.com/woocommerce/woocommerce-admin/pull/5762

**Comment by DanReyLop** [comment]
> I can't reproduce the error, but if you switch to this woo-flavoured Dependency Extraction Webpack Plugin does it help? It will handle WC dependencies as well as WP deps. I made that change in 431386c. Does it resolve the issue?

@psealock I don't see how that could be the issue. We aren't using the WC version of the Dependency Extraction plugin, but [we're manually doing more or less the same](https://github.com/Automattic/woocommerce-payments/blob/master/webpack.config.js#L65-L93).

The change you made didn't fix it for me.

I'm gonna try a couple of things and get back to you. I find it really puzzling that you can't reproduce it on your end :/

**Comment by psealock** [comment]
>I don't see how that could be the issue. We aren't using the WC version of the Dependency Extraction plugin, but we're manually doing more or less the same.

Oh right, I didn't see that. I will revert my commit then and it can be added separately. In my experience, getting these errors to show up comes down to loading order of other plugins or scripts with dependencies of their own. It could have been an error anywhere in the stack and simple luck prevented it from showing up until now.

One thing I did notice in `webpack.config.js` is there is no matching `requestToHandle` for WC Blocks. Could that be an issue?

https://github.com/Automattic/woocommerce-payments/blob/9440d8bbb5fb58dc7682b4e7193ddaf92e4a7fb7/webpack.config.js#L78

**Comment by DanReyLop** [comment]
> One thing I did notice in webpack.config.js is there is no matching requestToHandle for WC Blocks. Could that be an issue?

It's not an issue **right now**, but it's a future hard-to-track bug waiting to happen. Good eye! I'll add it (it didn't fix the current problem though).

> In my experience, getting these errors to show up comes down to loading order of other plugins or scripts with dependencies of their own.

Yeah, a loading order issue seems to be a good guess. Webpack is working as it should, `wc-navigation` is included in the array of deps for the `settings.js` [tiny script](https://github.com/Automattic/woocommerce-payments/blob/master/client/settings.js), and I can see that the `woocommerce-admin/dist/navigation/index.js` script is included in the page. I'll keep digging.

**Comment by DanReyLop** [comment]
Interesting... because `navigation/index.js` is huge (I built WC-Admin using `npm run dev` so it's not minified), by the time `settings.js` has finished downloading, `navigation/index.js` hasn't been parsed yet.

Does that make sense? The `settings.js` script calls `@woocommerce/navigation.getQuery()` when it's run, without waiting for any kind of event. I realize now that call is not really needed so I deleted it and now everything works, but that left me uneasy. Does everything rely on the dependencies of your script loading before it by pure chance???

I've pushed a "fix": https://github.com/Automattic/woocommerce-payments/pull/1136/commits/fd9d22157f7d8aafb785e81382faf3a3111a72aa

**Comment by psealock** [comment]
Here is another avenue to try. I just rebased this against `main` to include changes relevant to this PR. It uses DEWP to ensure dependencies are correct on wc-admin end. Can you try against this branch? https://github.com/woocommerce/woocommerce-admin/pull/5762

Incidentally, you'll need to have Gutenberg plugin built and activated to get the Nav working because `wp.components` is not bundled. If this fixes the problem, then I can work on it

File: client/index.js
Comment (psealock) [reply client/index.js:39]: Thanks @joshuatf, fixed in https://github.com/Automattic/woocommerce-payments/pull/1136/commits/39d45d6acaba165ddcad6a4e4fc08530a3b361e1

Comment (psealock) [reply client/index.js:40]: And added a better check in https://github.com/Automattic/woocommerce-payments/pull/1136/commits/26af034037debe38c3fdf47b78a657d037297d13

**Comment by psealock** [comment]
>Interesting... because navigation/index.js is huge (I built WC-Admin using npm run dev so it's not minified), by the time settings.js has finished downloading, navigation/index.js hasn't been parsed yet.

`wc-navigation` has a dependency on `wp-components` and because we are bundling `wp-components` that is likely where the bulk comes from. The effort to fix that is here https://github.com/woocommerce/woocommerce-admin/pull/5753

**Comment by DanReyLop** [comment]
> Here is another avenue to try. I just rebased this against main to include changes relevant to this PR. It uses DEWP to ensure dependencies are correct on wc-admin end. Can you try against this branch? woocommerce/woocommerce-admin#5762

It works! (Even after I added back the usage of `@woocommerce/navigation`). However, I think it's just because of this:

> wc-navigation has a dependency on wp-components and because we are bundling wp-components that is likely where the bulk comes from.

So, now that `navigation/index.js` weighs just ~100kb, it finishes loading before WCPay's `settings.js`. Before it weighted about 5MB. Again, now it works, but it makes me feel like this Webpack externals system is incredibly brittle. Not the fault of this PR, of course, just a thought.

**Comment by psealock** [comment]
> It works!

Ok, fantastic. It sounds like it works because of the reduced size in `wc-navigation` from unbundling `wc-components` as opposed to correct dependencies in wc-admin, but I'd need to check to confirm.

We are waiting on WP 5.6.0 to be released December 8 before moving forward with the unbundling PR. 

>Again, now it works, but it makes me feel like this Webpack externals system is incredibly brittle.

True, time to parse may not have been considered

**Review by DanReyLop — APPROVED** [review]
I don't know if you were waiting for a final approval, but this looks ready to merge, together with https://github.com/Automattic/woocommerce-payments/pull/1067. They work fine with current versions of WC and WC-Admin, so there's no point in waiting any longer. Do you agree, or are you holding out in case there are any last-minute changes to WC-Admin?

**Comment by psealock** [comment]
Thanks for the reviews and testing @DanReyLop ! The corresponding changes in wc-admin are merged and feels stable enough to go with this PR.

**Pull request merged**
